include/range/v3/utility/swap.hpp

```diff
@@ -42,10 +53,26 @@ namespace ranges
         /// \cond
         namespace adl_swap_detail
         {
-            using std::swap;
+
+            // The following implementation of swap is adapted from libc++:
+            template <class T>
+            RANGES_CXX14_CONSTEXPR
+            typename std::enable_if<
+                std::is_move_constructible<T>::value &&
+                std::is_move_assignable<T>::value
+             >::type
+            swap(T& x, T& y)
+            noexcept(std::is_nothrow_move_constructible<T>::value &&
+                     std::is_nothrow_move_assignable<T>::value)
+            {
+                T t(std::move(x));
+                x = std::move(y);
+                y = std::move(t);
+            }
```

Comment:
Thanks for taking a look at this. I didn't knew users could specialize `std::swap` in the `std` namespace. I thought the preferred way was to add a swap function to the namespace of the user-defined type and let ADL do the rest. If explicit specializations of `std::swap` are to be considered, then I need to reconsider this approach (more below).

First, what I want and tried to do is to let ADL pick the swap function in the namespace of the user-defined type. For literal types, or if no swap function is found, it should pick this swap function and not the one in namespace std. As far as I can tell, this overload does not create any ambiguity. It would if I were to import `std::swap`. 

If I have to allow for explicit specializations of std swap, then I need to import std::swap. I don't see a way to import those without importing `std::swap`. The whole thing is not doomed tho, since I can probably add a constexpr swap to the ranges namespace that will be picked for types living on it, and users can add constexpr swap functions for their own types if desired. However, I don't see a way to add a constexpr swap for C-arrays if I have to import the std::swap here (I've just realized that I forgot to add the swap overload for C-arrays that calls `swap_ranges` but I can fix this.). What do you think?
 
> std::move isn't constexpr in C++11.

The function is only constexpr in C++14 or greater.